bpo-32689: Updates shutil.move to allow for Path objects to be used as source arg

maxwellmckinnon opened the pull request:
Important work originally done by @emilyemorehouse two years ago and nearly ready to go in.

This bug has affected many people and in some cases has been a dealbreaker to the adoption of the otherwise wonderful pathlib and PEP519. https://stackoverflow.com/questions/33625931/copy-file-with-pathlib-in-python.

This adds the outstanding test request from that PR @vstinner (https://github.com/python/cpython/pull/5393).

Test fails without the change, passes with it, along with every other test in test_shutil.

Some variants were experimented with to make the one line change and the most performant one was picked.


# Added Test for PathLike directory destination, the current fail case

```
Lib/test/test_shutil.py::TestMove::test_move_file_pathlike FAILED                                                               [100%]

============================================================== FAILURES ===============================================================
__________________________________________________ TestMove.test_move_file_pathlike ___________________________________________________

self = <test.test_shutil.TestMove testMethod=test_move_file_pathlike>

    def test_move_file_pathlike(self):
        # Move a file to another location on the same filesystem.
        src = pathlib.Path(self.src_file)
>       self._check_move_file(src, self.dst_dir, self.dst_file)

Lib/test/test_shutil.py:1563:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
Lib/test/test_shutil.py:1545: in _check_move_file
    shutil.move(src, dst)
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/shutil.py:562: in move
    real_dst = os.path.join(dst, _basename(src))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

path = PosixPath('/var/folders/r2/psq74t5x3nbfzlph8bh2pvdw0000gn/T/tmp9ie0wh9_/foo')

    def _basename(path):
        # A basename() variant which first strips the trailing slash, if present.
        # Thus we always get the last component of the path, even for directories.
        sep = os.path.sep + (os.path.altsep or '')
>       return os.path.basename(path.rstrip(sep))
E       AttributeError: 'PosixPath' object has no attribute 'rstrip'

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/shutil.py:526: AttributeError
============================================== 1 failed, 102 deselected in 0.30 seconds ===============================================
```

After change:

```
========================================================= test session starts =========================================================
platform darwin -- Python 3.7.4, pytest-5.0.1, py-1.8.0, pluggy-0.12.0 -- /Users/maxwellmckinnon/.venvs/TA3.7/bin/python3.7
cachedir: .pytest_cache
rootdir: /Users/maxwellmckinnon/dev/cpython
plugins: cov-2.7.1, mock-1.10.4
collected 103 items / 102 deselected / 1 selected

Lib/test/test_shutil.py::TestMove::test_move_file_pathlike PASSED                                                               [100%]

============================================== 1 passed, 102 deselected in 0.06 seconds ===============================================
```

Running all the tests in test_shutil.py
```
╰─ pytest Lib/test/test_shutil.py -v
========================================================= test session starts =========================================================
platform darwin -- Python 3.7.4, pytest-5.0.1, py-1.8.0, pluggy-0.12.0 -- /Users/maxwellmckinnon/.venvs/TA3.7/bin/python3.7
cachedir: .pytest_cache
rootdir: /Users/maxwellmckinnon/dev/cpython
plugins: cov-2.7.1, mock-1.10.4
collected 103 items

Lib/test/test_shutil.py::TestShutil::test_chown PASSED                                                                          [  0%]
Lib/test/test_shutil.py::TestShutil::test_copy PASSED                                                                           [  1%]
...
Lib/test/test_shutil.py::TermsizeTests::test_stty_match SKIPPED                                                                 [ 99%]
Lib/test/test_shutil.py::PublicAPITests::test_module_all_attribute PASSED                                                       [100%]

================================================ 96 passed, 7 skipped in 1.25 seconds =================================================
```

# Performance Considerations
Is it considered poor form to get rid of _basename altogether and make use of pathlib in the move function? I'm not sure if the idea is for all these modules to strictly avoid circular dependencies. They are already using os.path which is just as much a citizen in 3.8 as pathlib right?

e.g.

`real_dst = os.path.join(dst, _basename(src))`
becomes
`real_dst = Path(dst) / Path(src).name`

I've looked around and familiarized myself, and I now think importing pathlib here is fine. My only remaining concern is that of performance.

Here's the performance difference for this step. 

```
In [46]: %timeit real_dst = os.path.join("a/b/c", _basename('b/'))
2.71 µs ± 62.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [47]: %timeit real_dst = Path("a/b/c") / Path('b/').name
12.4 µs ± 65.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Is 10us significant or insignificant compared to the least expensive operation this function will do? I don't know. Let's find out.

```
In [55]: %timeit os.rename('/tmp/a/a.txt', '/tmp/a/b.txt'); os.rename('/tmp/a/b.txt', '/tmp/a/a.txt')
124 µs ± 2.18 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
62us to rename. 10us seems significant enough that we wouldn't want to favor the Path sugar suggestion. 16% speed decrease from adding the 10us.

What do people think? I was hoping to get to use pathlib.Path here, but I suspect for this low level move, it should be as fast as possible, and 16% is not worth one line of sugary code to me.


<!-- issue-number: [bpo-32689](https://bugs.python.org/issue32689) -->
https://bugs.python.org/issue32689
<!-- /issue-number -->


Automerge-Triggered-By: @gvanrossum

emilyemorehouse committed:
bpo-32689: Updates shutil.move function to allow for Path objects to be used as source argument by casting to string before performing rstrip. Added documentation on why rstrip needs to be used in the helper _basename method.

emilyemorehouse committed:
Adding news entry

Commit:
Add test to show breakage for shutil.move where destination is a directory in the form of os.PathLike (pathlib.Path(dir)).

Commit:
Taking a guess at what the whitespace issue was.

Commit:
The results of running reindent.check on the file. Hope this fixes the whitespace gating issue.

brandtbucher commented on Lib/shutil.py:
```diff
@@ -550,7 +562,10 @@ def move(src, dst, copy_function=copy2):
+        # Using _basename instead of os.path.basename is important, as we must
+        # ignore any trailing slash to avoid the basename returning ''
         real_dst = os.path.join(dst, _basename(src))
+
```
I'd avoid introducing whitespace here.

```suggestion
```

brandtbucher commented on Lib/shutil.py:
```diff
@@ -511,8 +511,20 @@ def onerror(*args):
+    """A basename() variant which first strips the trailing slash, if present.
+    Thus we always get the last component of the path, even for directories.
+
+    path: Union[PathLike, str]
```
Maybe it's just me, but this docstring seems unnecessarily verbose for such a simple function. The wording is good enough that it's pretty clear what's happening without needing the examples below.

maxwellmckinnon replied:
No one mentioned this in the original PR this is following up on, but I don't have a strong opinion here. Either way is fine with me.

brandtbucher reviewed (changes requested):
A couple minor nitpicks, but otherwise looks good! `dst` is already pathlike-safe, correct?

I'd also add a note to the docs for this in `Doc/library/shutil.rst`:

```rst
.. versionchanged:: 3.9
      Accepts a :term:`path-like object` for both *src* and *dst*.
```

I also vote -1 on the `Path` slowdown, for the reasons you mentioned.

maxwellmckinnon committed:
remove whitespace

Will this fail the linter again?

Co-Authored-By: Brandt Bucher <brandtbucher@gmail.com>

Commit:
Update doc .rst files

Commit:
Merge branch 'fix-issue-32689' of https://github.com/maxwellmckinnon/cpython into fix-issue-32689

maxwellmckinnon committed:
Merge branch 'master' into fix-issue-32689

maxwellmckinnon committed:
Update shutil.py

Reverting the whitespace edit as the tests didn't pass

maxwellmckinnon committed:
Update shutil.rst

Fixed spacing failing doc building test

Commit:
ran python reindent.py -v ~/dev/cpython/Lib/shutil.py to fix whitespace break in CI

ammaraskar commented on Lib/test/test_shutil.py:
```diff
@@ -1825,6 +1825,11 @@ def test_move_file_to_dir(self):
 
+    def test_move_file_to_dir_pathlike(self):
+        # Move a file to another location on the same filesystem.
+        src = pathlib.Path(self.src_file)
```
Could you add a corresponding test for a `Path` object as the `dst`? It's important to keep it tested so no one breaks the behavior accidentally in the future.

maxwellmckinnon replied:
Yep it's not under test now and free to regress so great call.

Done and passing!

ammaraskar commented on Doc/library/shutil.rst:
```diff
@@ -352,6 +352,10 @@ Directory and files operations
+   .. versionchanged:: 3.9
+      Accepts a :term:`path-like object` for both *src* and *dst*.
+
+
```
Please follow the convention of the rest of the file and make this a single blank line.

ammaraskar commented on Misc/NEWS.d/next/Library/2018-02-13-17-58-30.bpo-32689.a-3SnH.rst:
```diff
@@ -0,0 +1,2 @@
+Updates shutil.move() function to allow for Path objects to be used as
```
```suggestion
Update :func:`shutil.move` function to allow for Path objects to be used as
```

ammaraskar reviewed (commented):
Just gonna ping Emily on the original PR to see if she wants to finish off the work herself.

maxwellmckinnon committed:
Update Misc/NEWS.d/next/Library/2018-02-13-17-58-30.bpo-32689.a-3SnH.rst

Co-Authored-By: Ammar Askar <ammar_askar@hotmail.com>

maxwellmckinnon committed:
Update shutil.rst

Remove whitespace line after versionchanged:: 3.9

maxwellmckinnon committed:
Update test_shutil.py

Added test for destination pathlike case as well (currently working but best to enforce the behavior!)

Commit:
Fixed another whitespace blocker I introduced via online editing -- squash this later -- ran Tools/scripts/reindent.py -v ~/dev/cpython/Lib/test/test_shutil.py

gvanrossum reviewed (approved):
OK, let this go into 3.9.
